wp-admin/js/widgets/media-widgets.js

```diff
@@ -303,38 +341,117 @@ wp.mediaWidgets = ( function( $ ) {
+		 * by PHP, where each widget update cause the entire element to be emptied
+		 * and replaced with the rendered output of WP_Widget::form() which is
+		 * sent back in Ajax request made to save/update the widget instance.
+		 * To prevent a "flash of replaced DOM elements and re-initialized JS
+		 * components", the JS template is rendered outside of the normal form
+		 * container.
+		 */
+		controlContainer = $( '<div class="media-widget-control"></div>' );
+		widgetContent.before( controlContainer );
+
+		/*
 		 * Sync the widget instance model attributes onto the hidden inputs that widgets currently use to store the state.
 		 * In the future, when widgets are JS-driven, the underlying widget instance data should be exposed as a model
 		 * from the start, without having to sync with hidden fields. See <https://core.trac.wordpress.org/ticket/33507>.
 		 */
-		widgetId = widgetForm.find( '> .widget-id' ).val();
-		controlContainer = $( '<div class="media-widget-control"></div>' );
-		widgetContent.before( controlContainer );
-		modelAttributes = {
-			id: widgetId
-		};
+		modelAttributes = {};
 		widgetContent.find( '.media-widget-instance-property' ).each( function() {
 			var input = $( this );
 			modelAttributes[ input.data( 'property' ) ] = input.val();
 		} );
+		modelAttributes.id = widgetId;
 
-		model = new ModelConstructor( modelAttributes );
+		widgetModel = new ModelConstructor( modelAttributes );
 
 		widgetControl = new ControlConstructor( {
 			el: controlContainer,
-			model: model
+			model: widgetModel
 		} );
 		widgetControl.render();
 
-		// @todo Sync the properties from the inputs into the model upon widget-synced and widget-updated?
-		// @todo There is no widget-removed event.
-		component.modelCollection.add( [ model ] );
+		/*
+		 * Note that the model and control currently won't ever get garbage-collected
+		 * when a widget gets removed/deleted because there is no widget-removed event.
+		 */
+		component.modelCollection.add( [ widgetModel ] );
+		component.widgetControls[ widgetModel.get( 'id' ) ] = widgetControl;
+	};
 
-		// @todo Register model?
-		component.widgetControls[ widgetId ] = widgetControl;
-	} );
+	/**
+	 * Sync widget instance data sanitized from server back onto widget model.
+	 *
+	 * This gets called via the 'widget-updated' event when saving a widget from
+	 * the widgets admin screen and also via the 'widget-synced' event when making
+	 * a change to a widget in the customizer.
+	 *
+	 * @param {jQuery.Event} event - Event.
+	 * @param {jQuery}       widgetContainer - Widget container element.
+	 * @returns {void}
+	 */
+	component.handleWidgetUpdated = function handleWidgetUpdated( event, widgetContainer ) {
+		var widgetForm, widgetContent, widgetId, widgetControl, attributes = {};
+		widgetForm = widgetContainer.find( '> .widget-inside > .form, > .widget-inside > form' );
+		widgetContent = widgetForm.find( '> .widget-content' );
```

Comment:
All of the hidden inputs for the instance properties will be inside of `.widget-content`, as it is what `WP_Media_Widget::form()` renders. I suppose we could just use `widgetForm` but it seemed like it would be better to restrict the scope of the selector query if we know they're only ever going to be inside that element.